### PR TITLE
Add support for finding `mod.sp` up `cwd` tree. Closes #1533

### LINKED ISSUE #1533: `mod.sp` resolution should traverse up the directory tree
* `check` and `dashboard` SHOULD fail if a valid `mod.sp` is not found
* `query` should continue and load up pseudo-resources, even if a `mod.sp` is not found

### PR DESCRIPTION
**Pull request by binaek**

### CODE REVIEW

File: tests/acceptance/test_files/022.mod-sp-traversal.bats [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  # go to the nested sub directory within the mod
+  cd $FILE_PATH/test_data/nested_mod/folder1/folder11/folder111
+
+  run steampipe check all
+  assert_success
+  cd -
+}
+
+# This test consists of a mod with nested folders, with mod.sp file within one of them(folder11).
+# Running steampipe check from folder1 should return an error, since the mod.sp file is present nowhere
```
Comment (kaidaguerre) [inline tests/acceptance/test_files/022.mod-sp-traversal.bats:17]: ```Running steampipe check from folder1 (i.e. _above_ the mod folder) should return an error, ...```

File: workspace/workspace_error.go [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package workspace
+
+import "errors"
+
+var (
+	ErrModSpNotFound = errors.New("this command requires a mod definition file - could not find in the current directory tree")
```
Comment (kaidaguerre) [inline workspace/workspace_error.go:6]: ErrorNoModDefinition

**Review by kaidaguerre — CHANGES_REQUESTED** [review]

File: tests/acceptance/test_files/022.mod-sp-traversal.bats (hunk 1 above)
Comment (pskrbasu) [reply tests/acceptance/test_files/022.mod-sp-traversal.bats]: updated comment

File: cmd/dashboard.go [hunk 3]
```diff
@@ -55,6 +57,18 @@ func runDashboardCmd(cmd *cobra.Command, args []string) {
 		logging.LogTime("runDashboardCmd end")
 		if r := recover(); r != nil {
 			utils.ShowError(dashboardCtx, helpers.ToError(r))
+			if viper.GetBool(constants.ArgServiceMode) {
+				state, _ := dashboardserver.GetDashboardServiceState()
```
Comment (kaidaguerre) [inline cmd/dashboard.go:61]: wrap in a function

File: cmd/dashboard.go [hunk 4]
```diff
@@ -83,7 +97,22 @@ func runDashboardCmd(cmd *cobra.Command, args []string) {
 
 	server.Start()
 
-	if !viper.GetBool(constants.ArgServiceMode) {
+	if viper.GetBool(constants.ArgServiceMode) {
+		state := &dashboardserver.DashboardServiceState{
```
Comment (kaidaguerre) [inline cmd/dashboard.go:101]: wrap in a function

File: constants/exit_codes.go [hunk 5]
```diff
@@ -0,0 +1,5 @@
+package constants
+
+const (
+	EC_NoModFile = 15
```
Comment (kaidaguerre) [inline constants/exit_codes.go:4]: So tell me - what am I going to say about this variable name?

Now try to ask yourself that question before submitting the PR :p 

`ExitCodeNoModFile` please

And if we are defining consts for exit codes (a good idea), please raise an issue to use them consistently for all commands

File: dashboard/dashboardserver/service.go [hunk 6]
```diff
@@ -108,56 +115,70 @@ func RunForService(ctx context.Context, serverListen ListenType, serverPort List
...
-	utils.LogTime("db.waitForConnection start")
-	defer utils.LogTime("db.waitForConnection end")
+// when started as a service, 'steampipe dashboard' always writes a
+// state file in 'internal' with the outcome - even on failures
+// this function polls for the file and loads up the error, if any
+func waitForDashboardService(ctx context.Context) error {
+	utils.LogTime("db.waitForDashboardServerStartup start")
+	defer utils.LogTime("db.waitForDashboardServerStartup end")
 
 	pingTimer := time.NewTicker(10 * time.Millisecond)
```
Comment (kaidaguerre) [inline dashboard/dashboardserver/service.go:128]: 10 ms seems a bit too fast - 25? 50?

File: dashboard/dashboardserver/service.go [hunk 7]
```diff
@@ -108,56 +115,70 @@ func RunForService(ctx context.Context, serverListen ListenType, serverPort List
...
-	timeoutAt := time.After(5 * time.Second)
 	defer pingTimer.Stop()
 
+	timeoutAt := time.After(5 * time.Second)
```
Comment (kaidaguerre) [inline dashboard/dashboardserver/service.go:131]: const for timeout please

File: dashboard/dashboardserver/service.go [hunk 8]
```diff
@@ -108,56 +115,70 @@ func RunForService(ctx context.Context, serverListen ListenType, serverPort List
...
+			return nil
 		case <-timeoutAt:
-			return fmt.Errorf("dashboard server startup failed")
+			return fmt.Errorf("dashboard server startup timed out")
 		}
 	}
 }
+
+func WriteServiceStateFile(state *DashboardServiceState) error {
+	stateBytes, err := json.Marshal(state)
```
Comment (kaidaguerre) [inline dashboard/dashboardserver/service.go:166]: lets use MarshalIndent

**Review by kaidaguerre — CHANGES_REQUESTED** [review]

File: dashboard/dashboardserver/service.go [hunk 9]
```diff
@@ -108,56 +115,69 @@ func RunForService(ctx context.Context, serverListen ListenType, serverPort List
...
+			state, err := loadServiceStateFile()
+			if err != nil {
+				if os.IsNotExist(err) {
+					// if the file hasn't been generated yet, that means 'dashboard' is still booting up
+					continue
+				}
+				// there was an unexpected error
 				return err
 			}
+			if len(state.Error) > 0 {
```
Comment (kaidaguerre) [inline dashboard/dashboardserver/service.go:148]: update comment - `// check the state file for an error`

File: dashboard/dashboardserver/service.go [hunk 10]
```diff
@@ -108,56 +115,69 @@ func RunForService(ctx context.Context, serverListen ListenType, serverPort List
...
+				// there was an error during start up
+				// remove the state file, since we don't need it anymore
+				os.Remove(filepaths.DashboardServiceStateFilePath())
+				// and return the error from the state file
+				return errors.New(state.Error)
+			}
+
+			// we loaded the state and there was no error -- aal-is-well
```
Comment (kaidaguerre) [inline dashboard/dashboardserver/service.go:156]: delete `-- aal-is-well`

File: workspace/workspace.go [hunk 11]
```diff
@@ -146,25 +147,56 @@ func (w *Workspace) Close() {
 	}
 }
 
+func (w *Workspace) ModfileExists() bool {
+	return len(w.modFileLocation) > 0
```
Comment (kaidaguerre) [inline workspace/workspace.go:151]: lets call this `modFilePath` for consistency

File: cmd/dashboard.go (hunk 3 above)
Comment (kaidaguerre) [reply cmd/dashboard.go:61]:
```
if isRunningAsService() {
    saveErrorToDashboardState()
}
```

File: cmd/dashboard.go (hunk 4 above)
Comment (kaidaguerre) [reply cmd/dashboard.go:101]:
```
if isRunningAsService() {
    saveDashboardState()
}
```

**Review by kaidaguerre — CHANGES_REQUESTED** [review]

File: workspace/workspace.go [hunk 12]
```diff
@@ -148,24 +148,24 @@ func (w *Workspace) Close() {
...
 
 func (w *Workspace) ModfileExists() bool {
-	return len(w.modFileLocation) > 0
+	return len(w.modFilePath) > 0
 }
 
 // check  whether the workspace contains a modfile
 // this will determine whether we load files recursively, and create pseudo resources for sql files
 func (w *Workspace) setModfileExists() {
 	modFile, err := w.findModFilePath(w.Path)
```
[the remark this answers is not in the document]
Comment (binaek) [reply workspace/workspace.go:157]: we do now!

https://github.com/turbot/steampipe/pull/1536/commits/897963707b701e608f6edea43445ec24cdc28700

**Pull request merged**